### IMPORTING LIBRARIES AND LOADING DATASET

In [84]:
# Importing required libraries

import numpy as np
import pandas as pd

In [85]:
# Loading dataset in the df

df = pd.read_csv('tested.csv')
print("\nData loaded successfully")


Data loaded successfully


### LET'S START WITH DATA CLEANING AND PREPROCESSING

In [86]:
print("Checking how our dataset looks\n")
df.head()

Checking how our dataset looks



,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,892,0,3,"Kelly, Mr. James",male,34.5,0,0,330911,7.8292,NaN,Q
1,893,1,3,"Wilkes, Mrs. James (Ellen Needs)",female,47.0,1,0,363272,7.0000,NaN,S
2,894,0,2,"Myles, Mr. Thomas Francis",male,62.0,0,0,240276,9.6875,NaN,Q
3,895,0,3,"Wirz, Mr. Albert",male,27.0,0,0,315154,8.6625,NaN,S
4,896,1,3,"Hirvonen, Mrs. Alexander (Helga E Lindqvist)",female,22.0,1,1,3101298,12.2875,NaN,S


In [87]:
print("The columns in our dataset is :\n")
df.columns

The columns in our dataset is :



Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='str')

In [88]:
print("lets look at the overview of the dataset :\n")
print(f"The shape of the data is {df.shape}\n")
print(df.info())


lets look at the overview of the dataset :

The shape of the data is (418, 12)

<class 'pandas.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  418 non-null    int64  
 1   Survived     418 non-null    int64  
 2   Pclass       418 non-null    int64  
 3   Name         418 non-null    str    
 4   Sex          418 non-null    str    
 5   Age          332 non-null    float64
 6   SibSp        418 non-null    int64  
 7   Parch        418 non-null    int64  
 8   Ticket       418 non-null    str    
 9   Fare         417 non-null    float64
 10  Cabin        91 non-null     str    
 11  Embarked     418 non-null    str    
dtypes: float64(2), int64(5), str(5)
memory usage: 56.1 KB
None


#### MANUAL REPORT :-
##### WE HAVE TOTAL 418 ROWS AND 12 COLUMNS
##### AGE, FARE, CABIN - THESE COLUMNS HAVE MISSING VALUES
##### AGE COLUMN HAS FLOAT64 DTYPE
##### TICKET COLUMN HAS STR DTYPE

In [89]:
# Exploring the dataset to notice any inconsistencies
df.tail(10)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
408,1300,1,3,"Riordan, Miss. Johanna Hannah""""",female,NaN,0,0,334915,7.7208,NaN,Q
409,1301,1,3,"Peacock, Miss. Treasteall",female,3.0,1,1,SOTON/O.Q. 3101315,13.7750,NaN,S
410,1302,1,3,"Naughton, Miss. Hannah",female,NaN,0,0,365237,7.7500,NaN,Q
411,1303,1,1,"Minahan, Mrs. William Edward (Lillian E Thorpe)",female,37.0,1,0,19928,90.0000,C78,Q
412,1304,1,3,"Henriksson, Miss. Jenny Lovisa",female,28.0,0,0,347086,7.7750,NaN,S
413,1305,0,3,"Spector, Mr. Woolf",male,NaN,0,0,A.5. 3236,8.0500,NaN,S
414,1306,1,1,"Oliva y Ocana, Dona. Fermina",female,39.0,0,0,PC 17758,108.9000,C105,C
415,1307,0,3,"Saether, Mr. Simon Sivertsen",male,38.5,0,0,SOTON/O.Q. 3101262,7.2500,NaN,S
416,1308,0,3,"Ware, Mr. Frederick",male,NaN,0,0,359309,8.0500,NaN,S
417,1309,0,3,"Peter, Master. Michael J",male,NaN,1,1,2668,22.3583,NaN,C


##### NAME & TICKET COLUMNS HAS INCONSISTENCIES WHICH NEEDS TO BE FIXED

### HANDLING MISSING VALUES

In [90]:
missing_vals = df[['Age', 'Cabin', 'Fare']].isnull().sum()

In [91]:
for i in missing_vals:
    print(f"The missing values are {i} and missing values percent is {(i/len(df)) * 100} ")

The missing values are 86 and missing values percent is 20.574162679425836 
The missing values are 327 and missing values percent is 78.22966507177034 
The missing values are 1 and missing values percent is 0.23923444976076555 


#### 1) THE FARE COLUMN HAS 0.2% MISSING VALUES THUS WE CAN REMOVE THE MISSING ROW

In [92]:
# Removed 1 missing val and saved in original dataset

df = df.dropna(subset=['Fare'])

In [93]:
df['Fare'].isna().sum()

0

#### 2) THE AGE COLUMN HAS 20% MISSING VALUES THUS WE CANNOT ENTIRELY REMOVE ROWS BUT FILL THE VALUES WITH MEAN OF AGE

In [94]:
# Filling the Age col with mean

mean_age = df['Age'].mean()
df['Age'] = df['Age'].fillna(mean_age)


In [95]:
df['Age'].isnull().sum()

0

In [96]:
df.shape

(417, 12)

#### 3) THE CABIN COLUMN HAS 78% MISSING VALS THUS CREATING FEATURE (HAS_CABIN) AND DROPING COLUMN

In [97]:
df['Has_Cabin'] = df['Cabin'].notnull().astype(int)

In [98]:
df = df.drop('Cabin', axis=1)

In [99]:
df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Embarked,Has_Cabin
0,892,0,3,"Kelly, Mr. James",male,34.5,0,0,330911,7.8292,Q,0
1,893,1,3,"Wilkes, Mrs. James (Ellen Needs)",female,47.0,1,0,363272,7.0000,S,0
2,894,0,2,"Myles, Mr. Thomas Francis",male,62.0,0,0,240276,9.6875,Q,0
3,895,0,3,"Wirz, Mr. Albert",male,27.0,0,0,315154,8.6625,S,0
4,896,1,3,"Hirvonen, Mrs. Alexander (Helga E Lindqvist)",female,22.0,1,1,3101298,12.2875,S,0


In [100]:
df['Has_Cabin'].value_counts()

Has_Cabin
0    326
1     91
Name: count, dtype: int64

##### THUS 326 PEOPLE DONT HAVE CABIN AND 91 PEOPLE HAS CABIN

In [101]:
df.isna().sum()

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Embarked       0
Has_Cabin      0
dtype: int64

### ALL MISSING VALUES ARE HANDLED
### NOW CHECKING DUPLICATE ROWS

In [102]:
print(df.duplicated().sum())
print("NO duplicate rows found")

0
NO duplicate rows found


### HANDLING INCONSISTENCIES
#### THE NAME AND TICKET COLUMN HAS INCONSISTENCY

In [ ]:
# starting with Name col

df['Name']

0                                  Kelly, Mr. James
1                  Wilkes, Mrs. James (Ellen Needs)
2                         Myles, Mr. Thomas Francis
3                                  Wirz, Mr. Albert
4      Hirvonen, Mrs. Alexander (Helga E Lindqvist)
                           ...                     
413                              Spector, Mr. Woolf
414                    Oliva y Ocana, Dona. Fermina
415                    Saether, Mr. Simon Sivertsen
416                             Ware, Mr. Frederick
417                        Peter, Master. Michael J
Name: Name, Length: 417, dtype: str

In [104]:
df[['Surname', 'Name']] = df['Name'].str.split(',', expand=True)

In [105]:
df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Embarked,Has_Cabin,Surname
0,892,0,3,Mr. James,male,34.5,0,0,330911,7.8292,Q,0,Kelly
1,893,1,3,Mrs. James (Ellen Needs),female,47.0,1,0,363272,7.0000,S,0,Wilkes
2,894,0,2,Mr. Thomas Francis,male,62.0,0,0,240276,9.6875,Q,0,Myles
3,895,0,3,Mr. Albert,male,27.0,0,0,315154,8.6625,S,0,Wirz
4,896,1,3,Mrs. Alexander (Helga E Lindqvist),female,22.0,1,1,3101298,12.2875,S,0,Hirvonen


In [106]:
df['Name'].head(20)

0                                           Mr. James
1                            Mrs. James (Ellen Needs)
2                                  Mr. Thomas Francis
3                                          Mr. Albert
4                  Mrs. Alexander (Helga E Lindqvist)
5                                    Mr. Johan Cervin
6                                          Miss. Kate
7                                  Mr. Albert Francis
8                    Mrs. Joseph (Sophie Halaut Easu)
9                                     Mr. John Samuel
10                                           Mr. Ylio
11                                Mr. Charles Cresson
12              Mrs. John Pillsbury (Nelle Stevenson)
13                                       Mr. Benjamin
14     Mrs. Herbert Fuller (Carrie Constance Toogood)
15                 Mrs. Sebastiano (Argenia Genovesi)
16                                         Mr. Daniel
17                                         Mr. Gerios
18                          

In [107]:
df['Title'] = df['Name'].str.extract(' ([A-Za-z]+)\.')

<>:1: SyntaxWarning: invalid escape sequence '\.'
<>:1: SyntaxWarning: invalid escape sequence '\.'
C:\Users\patil\AppData\Local\Temp\ipykernel_21496\3701537643.py:1: SyntaxWarning: invalid escape sequence '\.'
  df['Title'] = df['Name'].str.extract(' ([A-Za-z]+)\.')


In [108]:
df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Embarked,Has_Cabin,Surname,Title
0,892,0,3,Mr. James,male,34.5,0,0,330911,7.8292,Q,0,Kelly,Mr
1,893,1,3,Mrs. James (Ellen Needs),female,47.0,1,0,363272,7.0000,S,0,Wilkes,Mrs
2,894,0,2,Mr. Thomas Francis,male,62.0,0,0,240276,9.6875,Q,0,Myles,Mr
3,895,0,3,Mr. Albert,male,27.0,0,0,315154,8.6625,S,0,Wirz,Mr
4,896,1,3,Mrs. Alexander (Helga E Lindqvist),female,22.0,1,1,3101298,12.2875,S,0,Hirvonen,Mrs


In [109]:
df['Title'].value_counts()

Title
Mr        239
Miss       78
Mrs        72
Master     21
Col         2
Rev         2
Ms          1
Dr          1
Dona        1
Name: count, dtype: int64

In [110]:
df['Title'] = df['Title'].replace({
     'Mr': 'Mr',
    'Mister': 'Mr',
    'Mrs': 'Mrs',
    'Miss': 'Mrs',
    'Ms':'Mrs'
    })

In [111]:
df['Title'].value_counts()

Title
Mr        239
Mrs       151
Master     21
Col         2
Rev         2
Dr          1
Dona        1
Name: count, dtype: int64

In [152]:
df['Name'] = df['Name'].str.replace(r'\b(Mr|Mrs|Master|Col|Rev|Dr|Dona|Miss)\.\s*', '', regex=True)

In [153]:
df['Name'] = df["Name"].str.replace('"', '', regex=False)

In [154]:
df['Name'] = df["Name"].str.replace('(', '', regex=False)
df['Name'] = df["Name"].str.replace(')', '', regex=False)

In [155]:
df['Name'] = df['Name'].str.replace(' ', '', regex=False).str.strip()

In [156]:
print("Name col is cleaned")
df.tail(7)

Name col is cleaned


,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Embarked,Has_Cabin
411,1303,1,1,WilliamEdwardLillianEThorpe,female,37,1,0,19928,90.0000,Q,1
412,1304,1,3,JennyLovisa,female,28,0,0,347086,7.7750,S,0
413,1305,0,3,Woolf,male,30,0,0,53236,8.0500,S,0
414,1306,1,1,Fermina,female,39,0,0,17758,108.9000,C,1
415,1307,0,3,SimonSivertsen,male,38,0,0,3101262,7.2500,S,0
416,1308,0,3,Frederick,male,30,0,0,359309,8.0500,S,0
417,1309,0,3,MichaelJ,male,30,1,1,2668,22.3583,C,0


In [ ]:
# surnme and title cols are not needed 

df.drop(['Surname', 'Title'], axis=1, inplace=True)

In [147]:
print(df.columns)
print(f"\n The shape of the data is {df.shape}")

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Embarked', 'Has_Cabin'],
      dtype='str')

 The shape of the data is (417, 12)


In [148]:
# starting with the title col

df['Ticket'].value_counts()

Ticket
17608      5
113503     4
2343       4
31029      3
17483      3
          ..
53236      1
17758      1
3101262    1
359309     1
2668       1
Name: count, Length: 361, dtype: Int64

In [ ]:
df['Ticket'] = df['Ticket'].str.replace(r'[^0-9]', '', regex=True)

In [128]:
df['Ticket']

0       330911
1       363272
2       240276
3       315154
4      3101298
        ...   
413      53236
414      17758
415    3101262
416     359309
417       2668
Name: Ticket, Length: 417, dtype: str

### ALL INCONSISTENCIES HANDLED
### CHANGING THE DATATYPES

In [131]:
df['Ticket'] = df['Ticket'].astype('Int32')

In [136]:
df['Age'] = df['Age'].astype('int32')

In [137]:
df.info()

<class 'pandas.DataFrame'>
Index: 417 entries, 0 to 417
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  417 non-null    int64  
 1   Survived     417 non-null    int64  
 2   Pclass       417 non-null    int64  
 3   Name         417 non-null    str    
 4   Sex          417 non-null    str    
 5   Age          417 non-null    int32  
 6   SibSp        417 non-null    int64  
 7   Parch        417 non-null    int64  
 8   Ticket       417 non-null    Int32  
 9   Fare         417 non-null    float64
 10  Embarked     417 non-null    str    
 11  Has_Cabin    417 non-null    int32  
dtypes: Int32(1), float64(1), int32(2), int64(5), str(3)
memory usage: 46.4 KB


In [138]:
df.describe()

,PassengerId,Survived,Pclass,Age,SibSp,Parch,Ticket,Fare,Has_Cabin
count,417.000000,417.000000,417.000000,417.000000,417.000000,417.000000,417.0,417.000000,417.000000
mean,1100.635492,0.364508,2.263789,30.119904,0.448441,0.393285,452756.167866,35.627188,0.218225
std,120.923774,0.481870,0.842077,12.584522,0.897568,0.982419,2290657.910131,55.907576,0.413537
min,892.000000,0.000000,1.000000,0.000000,0.000000,0.000000,2.0,0.000000,0.000000
25%,996.000000,0.000000,1.000000,23.000000,0.000000,0.000000,13695.0,7.895800,0.000000
50%,1101.000000,0.000000,3.000000,30.000000,0.000000,0.000000,53337.0,14.454200,0.000000
75%,1205.000000,1.000000,3.000000,35.000000,1.000000,0.000000,345775.0,31.500000,0.000000
max,1309.000000,1.000000,3.000000,76.000000,8.000000,9.000000,23101291.0,512.329200,1.000000


### DTYPES CHANGED

In [140]:
cleaned_titanidata = df

In [ ]:
cleaned_titanidata.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Embarked,Has_Cabin
0,892,0,3,James,male,34,0,0,330911,7.8292,Q,0
1,893,1,3,JamesEllenNeeds,female,47,1,0,363272,7.0000,S,0
2,894,0,2,ThomasFrancis,male,62,0,0,240276,9.6875,Q,0
3,895,0,3,Albert,male,27,0,0,315154,8.6625,S,0
4,896,1,3,AlexanderHelgaELindqvist,female,22,1,1,3101298,12.2875,S,0
5,897,0,3,JohanCervin,male,14,0,0,7538,9.2250,S,0
6,898,1,3,Kate,female,30,0,0,330972,7.6292,Q,0
7,899,0,2,AlbertFrancis,male,26,1,1,248738,29.0000,S,0
8,900,1,3,JosephSophieHalautEasu,female,18,0,0,2657,7.2292,C,0
9,901,0,3,JohnSamuel,male,21,2,0,448871,24.1500,S,0
